File: HW7/7.1/7.1/list.cpp
#include <stdio.h>
#include "list.h"

struct Node
{
	int value;
	Node *next;
};

struct List
{
	Node *head = nullptr;
};

List *createList()
{
	return new List;
}

bool isEmpty(List *list)
{
	return list->head == nullptr;
}

void addValue(List *list, int value)
{
	if (isEmpty(list))
	{
		list->head = new Node {value, nullptr};
	}
	else
	{
		Node *first = list->head;
		Node *previous = nullptr;
		while (!isEmpty(list) && list->head->value < value)
		{
			previous = list->head;
			list->head = list->head->next;
		}
		if (previous != nullptr)
		{
			previous->next = new Node {value, list->head};
			list->head = first;
		}
		else
		{
			list->head = new Node {value, first};
		}
	}
}

bool deleteValue(List *list, int value)
{
	if (isEmpty(list))
	{
		return false;
	}
	Node *first = list->head;
	Node *previous = nullptr;
	while (!isEmpty(list) && list->head->value != value)
	{
		previous = list->head;
		list->head = list->head->next;
	}
	if (isEmpty(list))
	{
		list->head = first;
		return false;
	}
	if (previous != nullptr)
	{
		Node *temp = list->head->next;
		delete list->head;
		previous->next = temp;
		list->head = first;
	}
	else
	{
		Node *temp = list->head->next;
		delete list->head;
		list->head = temp;
	}
	return true;
}

void printList(List *list)
{
	Node *first = list->head;
	while (!isEmpty(list))
	{
		printf("%d ", list->head->value);
		list->head = list->head->next;
	}
	list->head = first;
}

void deleteList(List *list)
{
	while (!isEmpty(list))
	{
		Node *temp = list->head->next;
		delete list->head;
		list->head = temp;
	}
	delete list;
}

File: Semester2/Homeworks/HW3/Task2/Task2/List.cs
using System;

namespace Task2
{
    /// <summary>
    /// List, collection of string values.
    /// </summary>
    public class List
    {
        private Node head;

        /// <summary>
        /// List element, contains value and reference to next element.
        /// </summary>
        private class Node
        {
            /// <summary>
            /// Initializes a new instance of the <see cref="Node"/> class.
            /// </summary>
            /// <param name="value">Value of element</param>
            /// <param name="next">Reference to next element</param>
            public Node(string value, Node next)
            {
                Value = value;
                Next = next;
            }

            public string Value { get; set; }
            public Node Next { get; set; }
        }

        /// <summary>
        /// Checks for emptiness.
        /// </summary>
        /// <returns>True if the list is empty</returns>
        public bool IsEmpty => head == null;

        /// <summary>
        /// Adds a value to the beginning of the list.
        /// </summary>
        /// <param name="value">Value to add</param>
        public void AddToList(string value) => head = new Node(value, head);

        /// <summary>
        /// Deletes value from the list.
        /// </summary>
        /// <param name="value">Value for deletion</param>
        /// <returns>True if value was deleted</returns>
        public bool Delete(string value)
        {
            Node previous = null;
            var current = head;
            while (current != null)
            {
                if (Equals(current.Value, value))
                {
                    if (previous == null)
                    {
                        head = head.Next;
                        return true;
                    }
                    previous.Next = current.Next;
                    return true;
                }
                previous = current;
                current = current.Next;
            }
            return false;
        }

        /// <summary>
        /// Deletes and returns the first value in the list.
        /// </summary>
        /// <returns>The first value in the list</returns>
        /// <exception cref="InvalidOperationException">Thrown when the list is empty</exception>
        public string Delete()
        {
            if (IsEmpty)
            {
                throw new InvalidOperationException();
            }

            var value = head.Value;
            head = head.Next;
            return value;
        }

        /// <summary>
        /// Сhecks if the list contains a value.
        /// </summary>
        /// <param name="value">Value to check</param>
        /// <returns>True if the list contains value</returns>
        public bool Contains(string value)
        {
            var current = head;
            while (current != null)
            {
                if (Equals(current.Value, value))
                {
                    return true;
                }
                current = current.Next;
            }
            return false;
        }
    }
}
